instance/instance.go

```diff
@@ -140,147 +138,157 @@ func (s *Store) Add(w http.ResponseWriter, r *http.Request) {
 
 	log.Event(ctx, "add instance", log.INFO, logData)
 
-	b, err := func() ([]byte, error) {
-		instance, err := unmarshalInstance(ctx, r.Body, true)
-		if err != nil {
-			return nil, err
-		}
-
-		logData["instance_id"] = instance.InstanceID
+	instance, err := unmarshalInstance(ctx, r.Body, true)
+	if err != nil {
+		handleInstanceErr(ctx, err, w, logData)
+		return
+	}
 
-		instance.Links.Self = &models.LinkObject{
-			HRef: fmt.Sprintf("%s/instances/%s", s.Host, instance.InstanceID),
-		}
+	logData["instance_id"] = instance.InstanceID
 
-		instance, err = s.AddInstance(instance)
-		if err != nil {
-			log.Event(ctx, "add instance: store.AddInstance returned an error", log.ERROR, log.Error(err), logData)
-			return nil, err
-		}
+	instance.Links.Self = &models.LinkObject{
+		HRef: fmt.Sprintf("%s/instances/%s", s.Host, instance.InstanceID),
+	}
 
-		b, err := json.Marshal(instance)
-		if err != nil {
-			log.Event(ctx, "add instance: failed to marshal instance to json", log.ERROR, log.Error(err), logData)
-			return nil, err
-		}
+	instance, err = s.AddInstance(instance)
```

Comment:
Does this need wrapping with :
AcquireInstanceLock()
etc as done in dimension/dimension.go ... about line 45'ish ?

Reply:
In this case, we are creating a new instance, so it is impossible that there is a conflict, however we generate a new ETag and return it to the caller so that he/she can use it in later calls if required.